File: leetcode/226-invert-binary-tree/01.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def invertTree(self, root: TreeNode) -> TreeNode:
        if not root:
            return None
        stack = [root]
        while stack:
            node = stack.pop(0)
            node.left, node.right = node.right, node.left
            if node.left:
                stack.append(node.left)
            if node.right:
                stack.append(node.right)
        return root
  
    # a1 = TreeNode(4)
    # a2 = TreeNode(2)
    # a3 = TreeNode(7)
    # a1.left = a1
    # a1.right = a2

    # invertTree(a1)

    # print(a1.val)

File: leetcode/226-invert-binary-tree/01.js
function TreeNode(val) {
  this.val = val;
  this.left = null
  this.right = null
}

// var inverTree = function(root) {
//   subTree(root)
// }

// var subTree = function(node) {
//   if (node.left != null && node.right != null) {
//     var temp =   node.left
//     node.left = node.right
//     node.right = temp;
//     subTree(node.left)
//     subTree(node.right)
//   }  
// }


var inverTree = function(root) {
  if (!root) return root
  // const left = root.left
  // const right = root.right
  // root.right = inverTree(left)
  // root.left = inverTree(right)

  const stack = [root]
  let current = null;
  while((current = stack.pop())) {
    const left = current.left;
    const right = current.right;
    current.right = left
    current.left = right
    if (left) stack.push(left)
    if (right) stack.push(right)
  }

  return root
}


var t1 = new TreeNode(4)
var t2 = new TreeNode(2)
var t3 = new TreeNode(7)

t1.left = t2
t1.right = t3

inverTree(t1)

console.log(t1)
